feat(editor): Enhance Node Creator actions view

OlegIvaniv opened the pull request:
This PR refactors Node Creator so we can extend actions view. There are 3 main parts. 

### 1. Refactor NodeCreator to use new structure
The new design requires display of callouts and tooltips that are category-dependant.  This wouldn't be possible with the old flat structure because `ItemIterator` would simply render items based on their type. Instead we need a category-item container which renders the sub-items. To do this I've split the components into following structure:

<img width="1279" alt="CleanShot 2023-04-20 at 18 08 12@2x" src="https://user-images.githubusercontent.com/12657221/233424494-ddf77c48-b86d-4b20-a52e-e330cd242156.png">

This allows us to have more flexibility with rendering the items, and to conditionally render category content. 

### 2. Use SimplifiedNodeType for NodeCreator items

```
interface SimplifiedNodeType
	extends Pick<
		INodeTypeDescription,
		'displayName' | 'description' | 'name' | 'group' | 'icon' | 'iconUrl' | 'codex' | 'defaults'
	> {}
```
This PR also refactors NodeCreator to use SimplifiedNodeType. This is because we only need a subset of the properties from INodeTypeDescription to display nodes. And since we're rendering more than 300 items, the additional properties quickly add up to a lot of memory usage. 


### 3. Extract shared actions and state into composable
There's a lot of shared state and actions between NodeCreator components. This PR extracts them into a composable so we won't have to drill down props. These are:

#### useActions
Manages action-related tasks in the NodeCreator components. It handles locale-specific text retrieval, placeholder trigger creation, filtering and parsing of categorized action items, action data management and display conditions, appending manual trigger nodes when needed.

#### useActionsGeneration
Generates actions data based on node properties

#### useKeyboardNavigation
Handles the navigation state, key bindings, and their respective actions. It tracks watched keys, handles the navigation between selectable items, and supports registering custom hooks to handle specific key events.

#### useViewStacks
Handles shared state and actions for NodeCreator components by maintaining view stacks that represent different levels of navigation. It provides functionality to manipulate view stacks, manage active items, and handle search results within the stacks. The composable offers methods for pushing and popping view stacks, updating the current stack, and resetting the entire stack structure

![CleanShot 2023-04-20 at 17 29 09](https://user-images.githubusercontent.com/12657221/233425074-c5e755f3-60ff-452b-8a39-592d69ce9808.gif)

OlegIvaniv committed:
WIP

OlegIvaniv committed:
WIP

OlegIvaniv committed:
Extract actions into composable

OlegIvaniv committed:
WIP: Preserve categories when searching

OlegIvaniv committed:
WIP

OlegIvaniv committed:
Merge branch 'master' into ado-461-add-triggers-and-actions-top-level

OlegIvaniv committed:
WIP: Tweak styles

OlegIvaniv committed:
WIP: Refactor node creator

OlegIvaniv committed:
WIP: Finish Node Creator node view/subcategories refactor

OlegIvaniv committed:
WIP: Finished actions refactor

OlegIvaniv committed:
Cleanup & Lintfix

OlegIvaniv committed:
WIP: Improve memory managment

OlegIvaniv committed:
Fix interactions

OlegIvaniv committed:
WIP

OlegIvaniv committed:
WIP: Keyboard navigation

OlegIvaniv committed:
Improve keyboard navigation and memory managment

OlegIvaniv committed:
Finished view refactor

OlegIvaniv committed:
Merge branch 'master' into ado-461-add-triggers-and-actions-top-level

OlegIvaniv committed:
FIx custom api calls and activation callouts

OlegIvaniv committed:
Fix actions tracking and cleanup

OlegIvaniv committed:
Product review fixes

OlegIvaniv committed:
Telemetry fixes

OlegIvaniv committed:
Fix node creator e2es

OlegIvaniv committed:
Set action name font size and actionsEmpty font weight

OlegIvaniv committed:
Fix failing credentials spec

OlegIvaniv committed:
Make sure to select first action item when switching from nodes panel to actions panel

OlegIvaniv committed:
Add actions panel e2e tests

OlegIvaniv committed:
Merge branch 'master' into ado-461-add-triggers-and-actions-top-level

OlegIvaniv committed:
Cleanup

OlegIvaniv committed:
Fix actions generation and cleanup

OlegIvaniv committed:
Add correct Learn More link and adjust displaying of trigger icon

OlegIvaniv committed:
Change trigger icon condition to use nodeType group

OlegIvaniv committed:
Cleanup nodeTypesUtils and snapshots and lintfixes

OlegIvaniv committed:
Lint fixes

OlegIvaniv committed:
Merge branch 'master' into ado-461-add-triggers-and-actions-top-level

OlegIvaniv committed:
Refine logic to show trigger icon in node creator

OlegIvaniv committed:
Add unit tests & clean up

OlegIvaniv committed:
Add `003_auto_insert_action` experiment, hide empty sections for opposite root view

OlegIvaniv committed:
Lintfix

OlegIvaniv committed:
Merge branch 'master' into ado-461-add-triggers-and-actions-top-level

OlegIvaniv committed:
Do not show empty category tooltips and only show activation callout in triger root view

OlegIvaniv committed:
Fix no-results node creator view

OlegIvaniv committed:
Spacings tweaks and root rendering logic adjustment

OlegIvaniv committed:
Add unit tests

OlegIvaniv committed:
Merge branch 'master' into ado-461-add-triggers-and-actions-top-level

OlegIvaniv committed:
Lint and e2e fixes

OlegIvaniv committed:
Revert CLI changes, fix unit tests

OlegIvaniv committed:
Remove useless comments

OlegIvaniv committed:
Merge branch 'master' into ado-461-add-triggers-and-actions-top-level

OlegIvaniv committed:
Sync master, replace $externalHooks mixin

OlegIvaniv committed:
Lint fix

OlegIvaniv committed:
Focus first action when panel slides in, not category

mutdmour commented on packages/editor-ui/src/components/Node/NodeCreator/Modes/ActionsMode.vue:
```diff
@@ -0,0 +1,366 @@
+	const isPlaceholderTriggerAction = placeholderTriggerActions.some(
+		(p) => p.key === actionCreateElement.key,
+	);
+	// TODO: This should be controlled by A/B experiment
```
comment still needed?

mutdmour commented on packages/editor-ui/src/components/Node/NodeCreator/composables/useActions.ts:
```diff
@@ -0,0 +1,217 @@
+		return nodeTypes;
+	}
+
+	function setAddedNodeActionParameters(
```
can you explain to me what this function is doing? not sure I get it

OlegIvaniv replied:
It hooks into `addNode` action of workflow store. After node is added to the canvas, it sets the node's parameters based on the passed action data. I've added a small comment to explain

mutdmour replied:
ahhhhh that was my big confusion when I was looking at expanding this behevior recently.. would have expected the nodes panel to simply propogate the action to the canvas through events.. not a fan of this pattern because it's basically magic..  not suggesting we refactor this now.

mutdmour commented on packages/editor-ui/src/components/Node/NodeCreator/Panel/NodesListPanel.vue:
```diff
@@ -0,0 +1,257 @@
+					<p :class="$style.title" v-text="activeViewStack.title" v-if="activeViewStack.title" />
+				</div>
+				<p
+					v-if="activeViewStack.subtitle && !activeViewStack.hasHeaderBg"
```
do we need `hasHeaderBg`? could we just not pass `subtitle`?

OlegIvaniv replied:
Fair point, I've removed it and changed the `hasBg` classs condition to `!subtitle`.

mutdmour commented on packages/editor-ui/src/components/Node/NodeCreator/composables/useActions.ts:
```diff
@@ -0,0 +1,217 @@
+	}
+
+	function setAddedNodeActionParameters(
+		action: IUpdateInformation,
+		telemetry: Telemetry,
+		track = true,
+		rootView?: string,
```
could we simplify and not pass `telemetry` if not tracking?
```suggestion
		action: IUpdateInformation,
		telemetry?: Telemetry,
		rootView = '',
```

OlegIvaniv replied:
Good idea, updated

mutdmour commented on packages/editor-ui/src/components/Node/NodeCreator/Renderers/ItemsRenderer.vue:
```diff
@@ -0,0 +1,211 @@
+// Lazy render large items lists to prevent the browser from freezing
+// when loading many items.
+function renderItems() {
+	if (props.elements.length <= 20 || props.lazyRender === false) {
```
magic number.. make it a const and put it on top

mutdmour commented on packages/editor-ui/src/components/Node/NodeCreator/composables/useKeyboardNavigation.ts:
```diff
@@ -0,0 +1,156 @@
+	'Tab',
+];
+
+export const useKeyboardNavigation = defineStore('nodeCreatorKeyboardNavigation', () => {
```
😍

mutdmour reviewed (commented):
🚀

OlegIvaniv committed:
Address PR comments

OlegIvaniv commented on packages/editor-ui/src/components/Node/NodeCreator/ItemTypes/CategoryItem.vue:
```diff
@@ -0,0 +1,87 @@
+}
+
+:global(
+		[class*='_iteratorItem'][class*='_action_'] + [class*='_category_'] [class*='_categoryWrapper']
```
This rule is not necessary anymore, removed

OlegIvaniv commented on packages/editor-ui/src/components/Node/NodeCreator/ItemTypes/LabelItem.vue:
```diff
@@ -0,0 +1,36 @@
+}
+
+// Reduce top padding of the first label in a category
+:global([class*='_iteratorItem'][class*='_category_'] + [class*='_label_'] [class*='_label']) {
```
This rule is not necessary anymore, removed

OlegIvaniv commented on packages/editor-ui/src/Interface.ts:
```diff
@@ -25,7 +25,6 @@ import {
 	INodeCredentials,
 	INodeListSearchItems,
 	NodeParameterValueType,
-	INodeActionTypeDescription,
```
Yup, deleted

OlegIvaniv commented on packages/design-system/src/components/N8nCallout/Callout.vue:
```diff
@@ -64,24 +75,29 @@ export default defineComponent({
-	color: var(--color-info);
+	border-color: var(--callout-border-color, var(--color-foreground-base));
+	background-color: var(--callout-background-color, var(--color-background-light));
+	color: var(--callout-color, var(--color-info));
```
These are not needed anymore anyway, so I reverted these changes

OlegIvaniv commented on packages/editor-ui/src/components/Node/NodeCreator/composables/useActions.ts:
```diff
@@ -0,0 +1,217 @@
+				after(() => {
+					setLastNodeParameters(action);
+					if (track) trackActionSelected(action, telemetry, rootView || '');
+					storeWatcher();
```
Becasue we want to do this action only once. Calling this var will remove the subscription

OlegIvaniv commented on packages/editor-ui/src/components/Node/NodeCreator/utils.ts:
```diff
@@ -0,0 +1,72 @@
+	const createElement = {
+		uuid: uuidv4(),
+		key: node.name,
+		subcategory: subcategory ?? node.codex?.subcategories?.[CORE_NODES_CATEGORY]?.[0] ?? '*',
```
Good point. Replaced with the suggested const

OlegIvaniv committed:
Lint fix

OlegIvaniv commented on packages/editor-ui/src/utils/sortUtils.ts:
```diff
@@ -212,7 +214,74 @@ function getValue<T extends object>(obj: T, prop: string): unknown {
+	return results;
+}
+
+export function sublimeSearchSlow<T extends object>(
```
I reverted it to the original implementation, I was testing the efficiency of this method using for loop but the performance gains are minimal if any

OlegIvaniv commented on packages/editor-ui/src/plugins/i18n/locales/en.json:
```diff
@@ -739,6 +749,7 @@
 	"nodeCreator.categoryNames.sales": "Sales",
 	"nodeCreator.categoryNames.suggestedNodes": "Suggested Nodes ✨",
```
True. I removed the unused node creator locales and constants

OlegIvaniv committed:
Remove `setAddedNodeActionParameters` optional track param

OlegIvaniv committed:
Further simplify setAddedNodeActionParameters

OlegIvaniv committed:
Merge branch 'master' into ado-461-add-triggers-and-actions-top-level

# Conflicts:
#	packages/editor-ui/src/Interface.ts
#	packages/editor-ui/src/components/Node/NodeCreator/CategorizedItems.vue
#	packages/editor-ui/src/components/Node/NodeCreator/CategoryItem.vue
#	packages/editor-ui/src/components/Node/NodeCreator/ItemIterator.vue
#	packages/editor-ui/src/components/Node/NodeCreator/ItemTypes/ActionItem.vue
#	packages/editor-ui/src/components/Node/NodeCreator/ItemTypes/NodeItem.vue
#	packages/editor-ui/src/components/Node/NodeCreator/ItemTypes/SubcategoryItem.vue
#	packages/editor-ui/src/components/Node/NodeCreator/MainPanel.vue
#	packages/editor-ui/src/components/Node/NodeCreator/Panel/SearchBar.vue
#	packages/editor-ui/src/stores/nodeCreator.ts
#	packages/editor-ui/src/stores/nodeTypes.ts
#	packages/editor-ui/src/utils/nodeTypesUtils.ts
#	packages/editor-ui/src/views/NodeView.vue

OlegIvaniv committed:
Fix pnpn lock file

OlegIvaniv committed:
Fix types imports

OlegIvaniv commented:
@mutdmour Appreciate the thorough review! I addressed all your comments and fixed the two issues you've reported in Notion. Can you please have another look?

OlegIvaniv committed:
Merge branch 'master' into ado-461-add-triggers-and-actions-top-level

OlegIvaniv committed:
Fix 13-pinning spec

janober commented:
Got released with [`n8n@0.226.1`](https://github.com/n8n-io/n8n/releases/tag/n8n%400.226.1)
